[OCPBUGS-5808] Recreate k8s cloud.conf docs for ShiftStack (#55688)
openshift/openshift-docs: modules/cloud-conf-shiftstack-reference.adoc

//  Module included in the following assemblies:

//  * installing/installing_openstack/installing-openstack-cloud-config-reference.adoc

:_content-type: REFERENCE
[id="cloud-conf-shiftstack-reference_{context}"]
= OpenStack cloud provider options 

> The cloud provider configuration, typically stored as a file named `cloud.conf` controls how OpenShift Container Platform interacts with Red Hat OpenStack Platform (RHOSP).

Reviewer: ```suggestion
The cloud provider configuration, typically stored as a file named `cloud.conf`, controls how {product-title} interacts with Red Hat OpenStack Platform (RHOSP).
```

== Global options

The following options are used for {rh-openstack} CCM authentication with the {rh-openstack} Identity service, also known as Keystone. They are similiar to the global options that you can set by using the `openstack` CLI.

|===
| Option | Description

| `auth-url`
| The {rh-openstack} Identity service URL. For example, `http://128.110.154.166/identity`.

| `ca-file`
| Optional. The CA certificate bundle file for communication with the {rh-openstack} Identity service. If you use the HTTPS protocol with The Identity service URL, this option is required.

| `domain-id`
| The Identity service user domain ID. 

 Leave this option unset if you are using Identity service application credentials.

| `domain-name`
| The Identity service user domain name. 

 This option is not required if you set `domain-id`.

| `tenant-id`
| The Identity service project ID. Leave this option unset if you are using Identity service application credentials.

In version 3 of the Identity API, which changed the identifier `tenant` to `project`, the value of `tenant-id` is automatically mapped to the project construct in the API. 

| `tenant-name`
| The Identity service project name. 

> 

Reviewer: Extra return?

| `username`
| The Identity service user name. 

Leave this option unset if you are using Identity service application credentials.

| `password`
| The Identity service user password. 

 Leave this option unset if you are using Identity service application credentials.

| `region`
| The Identity service region name.

| `trust-id`
| The Identity service trust ID. A trust represents the authorization of a user, or trustor, to delegate roles to another user, or trustee. Optionally, a trust authorizes the trustee to impersonate the trustor. You can find available trusts by querying the `/v3/OS-TRUST/trusts` endpoint of the Identity service API.
|===

[id="ccm-config-lb-options"]

> == Load balancer options

Author: @dulek @MaysaMacedo Any objections to adding this to 4.12 only?

> == Load balancer options

Reviewer 2: Yes, this is 4.12+ only, this option is missing without external cloud provider.

The cloud provider supports several load balancer options for deployments that use Octavia. 

|===
| Option | Description

| `use-octavia`

> | Whether or not to use Octavia for the `LoadBalancer` type of service implementation rather than Neutron-LBaaS. The default value is `true`.

Reviewer: ```suggestion
| Whether or not to use Octavia for the `LoadBalancer` type of the service implementation rather than Neutron-LBaaS. The default value is `true`.
```

| `floating-network-id`
| Optional. The external network used to create floating IP addresses for load balancer virtual IP addresses (VIPs). If there are multiple external networks in the cloud, this option must be set or the user must specify `loadbalancer.openstack.org/floating-network-id` in the service annotation.

> 

Reviewer: Extra return?

| `lb-version`
| Optional. The load balancer API version. Only `"v2"` is supported.

| `subnet-id`
| The ID of the Networking service subnet on which load balancer VIPs are created. 

| `create-monitor`
| Whether or not to create a health monitor for the service load balancer. A health monitor is required for services that declare `externalTrafficPolicy: Local`. The default value is `false`.

This option is unsupported if you use {rh-openstack} earlier than version 17 with the `ovn` provider.

| `monitor-delay`
| The interval in seconds by which probes are sent to members of the load balancer. The default value is `5`.

| `monitor-max-retries`
| The number of successful checks that are required to change the operating status of a load balancer member to `ONLINE`. The valid range is `1` to `10`, and the default value is `1`.

| `monitor-timeout`
| The time in seconds that a monitor waits to connect to the back end before it times out. The default value is `3`.

|===

[id="ccm-config-metadata-options"]
== Metadata options

|===
| Option | Description

| `search-order`

> | This configuration key affects the way that the provider retrieves metadata that relates to the instances in which it runs. The default value of `configDrive,metadataService` results in the provider retrieving metadata that relates to the instance from, if available, the config drive first,and then the metadata service. Alternative values are:

Reviewer: ```suggestion
| This configuration key affects the way that the provider retrieves metadata that relates to the instances in which it runs. The default value of `configDrive,metadataService` results in the provider retrieving instance metadata from the configuration drive first if available, and then the metadata service. Alternative values are:
```
Purely a suggestion as this was how you explained the behavior of `metadataService,configDrive`.

| `lb-method`
| The load balancing algorithm used to create the load balancer pool.
For the Amphora provider the value can be `ROUND_ROBIN`, `LEAST_CONNECTIONS`, or `SOURCE_IP`. The default value is `ROUND_ROBIN`.

For the OVN provider, only the `SOURCE_IP_PORT` algorithm is supported.

For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `create-monitor` option as `true`  in the `cloud-provider-config` config map on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.

| `lb-provider`
| Optional. Used to specify the provider of the load balancer, for example, `amphora` or `octavia`. Only the Amphora and Octavia providers are supported.

| `lb-version`
| Optional. The load balancer API version. Only `"v2"` is supported.

| `subnet-id`
| The ID of the Networking service subnet on which load balancer VIPs are created. 

> This ID is also used to create pool members if `member-subnet-id` is not set.

Reviewer 2: We're mentioning an option that doesn't exist in this version.

Author: ```suggestion
```

TY!

> //  Module included in the following assemblies:

Author: @MaysaMacedo @stephenfin I'm trying to figure out which cloud conf options are supported in 4.9 (hopefully extending into 4.11) here. Any immediate thoughts?

Reviewer 2: The set of options that cloud provider accepts is the same in 4.8-4.11. The question might be what we actually support.

//  * installing/installing_openstack/installing-openstack-cloud-config-reference.adoc

:_content-type: REFERENCE
[id="cloud-conf-shiftstack-reference_{context}"]
= OpenStack cloud provider options 

The cloud provider configuration, typically stored as a file named `cloud.conf` controls how OpenShift Container Platform interacts with Red Hat OpenStack Platform (RHOSP).

You can create a valid `cloud.conf` file by specifying the following options in it.

[id="ccm-config-global-options"]
== Global options

The following options are used for {rh-openstack} CCM authentication with the {rh-openstack} Identity service, also known as Keystone. They are similiar to the global options that you can set by using the `openstack` CLI.

|===
| Option | Description

| `application-credential-id`
| The ID of an application credential to authenticate with. An `application-credential-secret` must be set along with this parameter.

| `application-credential-name`
| The name of an application credential to authenticate with. If `application-credential-id` is not set, the user name and domain must be set.

> | `application-credential-secret`

Reviewer 3: application credentials is not supported before 4.12, so application-credential-id, application-credential-name and application-credential-secret can be removed

| The secret of an application credential to authenticate with.

| `auth-url`
| The {rh-openstack} Identity service URL. For example, `http://128.110.154.166/identity`.

| `ca-file`
| Optional. The CA certificate bundle file for communication with the {rh-openstack} Identity service. If you use the HTTPS protocol with The Identity service URL, this option is required.

| `cert-file`
| Optional. The client certificate path to use for client TLS authentication.

| `key-file`

> | Optional. The client private key path to use for client TLS authentication.

Reviewer 2: These two will probably not work. Context: https://github.com/kubernetes/legacy-cloud-providers/blob/release-1.23/openstack/openstack.go#L158-L173

| `domain-id`
| The Identity service user domain ID. 

 Leave this option unset if you are using Identity service application credentials.

| `domain-name`
| The Identity service user domain name. 

 This option is not required if you set `domain-id`.

 | `os-endpoint-type`

> | The type of endpoint to use from the service catalog. 

Reviewer 2: Won't work.

If unset, public endpoints are used.

| `tenant-id`
| The Identity service project ID. Leave this option unset if you are using Identity service application credentials.

In version 3 of the Identity API, which changed the identifier `tenant` to `project`, the value of `tenant-id` is automatically mapped to the project construct in the API. 

| `tenant-name`
| The Identity service project name. 

| `tenant-domain-id`
| The Identity service project domain ID.

| `tenant-domain-name`
| The Identity service project domain name.

| `user-domain-id`
| The Identity service user domain ID.

| `user-domain-name`

> | The Identity service user domain name.

Reviewer 2: These are most likely missing too.

| `username`
| The Identity service user name. 

Leave this option unset if you are using Identity service application credentials.

| `password`
| The Identity service user password. 

 Leave this option unset if you are using Identity service application credentials.

| `region`
| The Identity service region name.

| `tls-insecure`

> | Whether or not to verify the server's TLS certificate. If set to `true`, the certificate is not verified. By default, the certificate is verified.

Reviewer 2: I don't see this one.

| `trust-id`
| The Identity service trust ID. A trust represents the authorization of a user, or trustor, to delegate roles to another user, or trustee. Optionally, a trust authorizes the trustee to impersonate the trustor. You can find available trusts by querying the `/v3/OS-TRUST/trusts` endpoint of the Identity service API.

| `trustee-id`
| The Identity service trustee user ID.

| `trustee-password`

> | The Identity service trustee user password.

Reviewer 2: These are missing too.

|===

[id="ccm-config-networking-options"]

> == Networking options

Reviewer 3: Networking options are also not supported

| `address-sort-order`
| This configuration key affects how the provider reports node addresses to Kubernetes node resources. The default order depends on the hard-coded order in which the provider queries addresses and what the cloud returns. A specific order is not guaranteed.

To override this behavior, specify a comma-separated list of CIDR addresses. The cloud provider sorts and groups all addresses that match the list in a prioritized manner, wherein the first retrieved item has  a higher priority than the last. Addresses that do not match the list remain in their default order. The default value is `""`.

This option can be useful if you have multiple or dual-stack interfaces attached to a node that need a user-controlled, deterministic way of sorting addresses.
|===

[id="ccm-config-lb-options"]
== Load balancer options

The cloud provider supports several load balancer options for deployments that use Octavia. 

|===
| Option | Description

| `enabled`

> | Whether or not to enable the `LoadBalancer` type of services integration. The default value is `true`.

Reviewer 2: This is missing too.

| `floating-subnet-id`
| Optional. The external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet-id`.

| `floating-subnet`
| Optional. A name pattern (glob or regular expression if starting with `~`) for the external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet`. If multiple subnets match the pattern, the first one with available IP addresses is used.

| `floating-subnet-tags`
| Optional. Tags for the external network subnet used to create floating IP addresses for the load balancer VIP. Can be overridden by the service annotation `loadbalancer.openstack.org/floating-subnet-tags`. If multiple subnets match these tags, the first one with available IP addresses is used.

> If the {rh-openstack} network is configured with sharing disabled, for example, with the `--no-share` flag used during creation, this option is unsupported. Set the network to share to use this option.

Reviewer 2: Doesn't exist in legacy cloud provider.

For the OVN provider, only the `SOURCE_IP_PORT` algorithm is supported.

For the Amphora provider, if using the `LEAST_CONNECTIONS` or `SOURCE_IP` methods, configure the `create-monitor` option as `true`  in the `cloud-provider-config` config map on the `openshift-config` namespace and `ETP:Local` on the load-balancer type service to allow balancing algorithm enforcement in the client to service endpoint connections.

| `lb-provider`
| Optional. Used to specify the provider of the load balancer, for example, `amphora` or `octavia`. Only the Amphora and Octavia providers are supported.

| `lb-version`
| Optional. The load balancer API version. Only `"v2"` is supported.

| `subnet-id`
| The ID of the Networking service subnet on which load balancer VIPs are created. 

This ID is also used to create pool members if `member-subnet-id` is not set.

| `member-subnet-id`
| ID of the Neutron network on which to create the members of the load balancer. The load balancer gets another network port on this subnet. Defaults to `subnet-id` if not set.

| `network-id`

> | The ID of the Networking service network on which load balancer VIPs are created. Unnecessary if `subnet-id` is set.

Reviewer 2: These two are missing too.

| `manage-security-groups`

> | If the Neutron security groups should be managed separately. Default: false

Reviewer 2: This exists, but I don't think we support it.

 This option is currently a workaround for the issue https://github.com/kubernetes/ingress-nginx/issues/3996, should be removed or refactored after the Kubernetes [KEP-1860](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1860-kube-proxy-IP-node-binding) is implemented.

| `default-tls-container-ref`
| Reference to a tls container. This option works with Octavia, when this option is set then the cloud provider will create an Octavia Listener of type TERMINATED_HTTPS for a TLS Terminated loadbalancer.

 Format for tls container ref: `https://{keymanager_host}/v1/containers/{uuid}`
 Check `container-store` parameter if you want to disable validation.

| `container-store`
| Optional. Used to specify the store of the tls-container-ref, e.g. "barbican" or "external" - other store will cause a warning log. Default value - `barbican` - existence of tls container ref would always be performed. If set to `external` format for tls container ref will not be validated.

| `max-shared-lb`
| The maximum number of services that can share a load balancer. The default value is `2`.
|===

> 

Reviewer 2: These should be removed too.
